### PR TITLE
Stabilize harmonic angles

### PR DESCRIPTION
**Pull request by proteneer**
This PR:

1) Completely removes the usage of GROMOS harmonic cosine angles and defaults back to the harmonic angle
2) Stabilizes the harmonic angle energy and forces, using Kahan's trick to greatly improve 32-bit precision (see plot below)
3) Verifies that nitriles now behave as intended, with the average equilibrium value closer to 3 rads as opposed to 2.8 rads. 
4) Modifies the stable angle to be trignometrically correct by moving the epsilon to an extra dimension of the input vectors. Before, as written, `arccos` and `atan2` implementations differed in the result. Now they are consistent.

image

TODO
-------
[ ] - Test alchemical protocol efficiency morphing a nitrile to a water.

### CODE REVIEW

**Review by jkausrelay — APPROVED** [review]

File: tests/nonbonded/test_nonbonded_mol_energy.py [hunk 1]
```diff
@@ -245,7 +245,7 @@ def test_nonbonded_mol_energy_matches_exchange_mover_batch_U_in_complex(precisio
         bound_impls,
         movers=[baro_impl],
     )
-    ctxt.multiple_steps(1000)
+    ctxt.multiple_steps(5000)
```
Comment (proteneer) [inline tests/nonbonded/test_nonbonded_mol_energy.py:248]: (This test was failing - after some offline discussion with @badisa  we're going to temporarily hack around this to unblock the PR and do a more systematic review of our energy tests later on).

File: timemachine/cpp/src/kernels/k_harmonic_angle.cuh [hunk 2]
```diff
@@ -19,67 +21,117 @@ void __global__ k_harmonic_angle(
...
-    RealType nijk = nij * njk;
-    RealType n3ij = nij * nij * nij;
-    RealType n3jk = njk * njk * njk;
+    RealType hi = 0;
+    RealType lo = 0;
+
+    // second pass, compute the hi/lo values
+    for (int d = 0; d < 3; d++) {
+        RealType vji = rji[d]; // coords[i_idx * D + d] - coords[j_idx * D + d];
+        RealType vjk = rjk[d]; // coords[k_idx * D + d] - coords[j_idx * D + d];
```
Comment (maxentile) [inline timemachine/cpp/src/kernels/k_harmonic_angle.cuh:57]: nit, now:
```suggestion
        RealType vji = rji[d]; // coords[i_idx * 3 + d] - coords[j_idx * 3 + d];
        RealType vjk = rjk[d]; // coords[k_idx * 3 + d] - coords[j_idx * 3 + d];
```

**Review by maxentile — APPROVED** [review]
This looks great! Seems like a clear improvement. Based on the 32-bit plots, I wonder if this removes one of the barriers to reducing coordinate precision…

Two comments:
1. Background — Is 2 atan2(y, x) commonly called “Kahan’s trick” ?
   - This method is preferred over alternatives in sec. 12 of [Kahan, 2006] “How Futile are Mindless Assessments of Roundoff in Floating-Point Computation?” , as you mentioned offline. May be useful to capture a ref in the `kahan_angle` docstring.
      - Additional pointer: Kahan’s suggestion is referenced and expanded in a nice blog post comparing precision loss among 5 ways to compute angles https://possiblywrong.wordpress.com/2020/07/17/computing-the-angle-between-two-vectors/ 
   - atan2 is sometimes used without comment for computing angles between vectors (e.g. in PBR4e, in TM periodic_torsions, in espaloma, and in a few Julia packages), although I’ve seen acos much more often (including in OpenMM, MDTraj, NNPOps, HOOMD, … ).
   - In previous discussions about the cos_angles part of the TM functional form, @jchodera has mentioned [[Swope, Ferguson, 1992]](https://www.semanticscholar.org/paper/Alternative-expressions-for-energies-and-forces-due-Swope-Ferguson/b0a8f125649de4e65047b485ff91154437af4a1f), which contains diagnosis of related numerical issues + some alternatives.
      - (Cross-linking two OpenMM reports of related numerical instabilities https://github.com/openmm/openmm/issues/1672 , https://github.com/openmm/openmm/issues/3515 .)

2. FE integration — This PR, or later?
   - Process uncertainty : Shall we merge this functional form change first, and figure out SingleTopology integration + any impact on computed free energies later?
   - Unsure how much of the discrepancy between TM and OpenMM for computing OFF2’s predictions on Freesolv is explained by cos_angles vs. other factors. Guess: very little.
   - Unsure if significant or minor changes to `interpolate_harmonic_angle_params` https://github.com/proteneer/timemachine/blob/457318a6d78da0affecdd270469f74d858ef0ac5/timemachine/fe/single_topology.py#L704 will be necessary. Guess: should be fine. Initially, the TODO in the PR description (interpolating nitrile <-> water) would be informative.
   - I think this change obviates a few investigations @jkausrelay and I briefly attempted in the past couple years to convert OpenFF angle parameters to TM angle parameters, but which stalled in validation. I do not have a good estimate for how much of an impact this change to the MD engine / functional form will have on FE predictions, but I suspect it may impact a large number of results by a small amount.

----------------

Closes https://github.com/proteneer/timemachine/issues/743 .

**Comment by proteneer** [comment]
> Is 2 atan2(y, x) commonly called “Kahan’s trick”

`atan2(y,x)` itself is not the Kahan's trick, as it's a well known primitive and documented extensively. You can use it directly:

`angle = atan2(vector2.y, vector2.x) - atan2(vector1.y, vector1.x);` (https://stackoverflow.com/questions/21483999/using-atan2-to-find-angle-between-two-vectors)

The real trick is in how `y` and `x` are formulated from the input vectors `rji, rjk` with an extra bit of normalization to improve precision:

```
nji = jnp.linalg.norm(rji, axis=-1, keepdims=True)
njk = jnp.linalg.norm(rjk, axis=-1, keepdims=True)
y = jnp.linalg.norm(njk * rji - nji * rjk, axis=-1)
x = jnp.linalg.norm(njk * rji + nji * rjk, axis=-1)
```

> In previous discussions about the cos_angles part of the TM functional form, @jchodera has mentioned [[Swope, Ferguson, 1992]](https://www.semanticscholar.org/paper/Alternative-expressions-for-energies-and-forces-due-Swope-Ferguson/b0a8f125649de4e65047b485ff91154437af4a1f), which contains diagnosis of related numerical issues + some alternatives.

Thanks for the link. There are *two* issues with the harmonic angle terms in the links 

1) With bond lengths that are non-zero for linear molecules, there's purely a numerical stability issue associated with angles near pi. This is analogous to the issue of computing small angles (which Kahan addresses).
2) With bond lengths that are close to zero (eg. in alchemical intermediates), there's a real singularity associated with the energies. This stems from the observation that a small perturbation, when two points are close to each other, can drastically affect the actual value of the angle (and the energy).

Kahan's trick primarily addresses 1), but also impacts 2) by changing the way epsilon is used. 

> FE integration — This PR, or later?

Later - I'm going to run some systematic tests (specifically with nitriles, allenes, etc.) before concluding anything on the impact of SingleTopology calcs.

> Initially, the TODO in the PR description (interpolating nitrile <-> water) would be informative.

I checked this - and overlaps etc. looked fine. But wasn't sure how to write a unit test for it.

File: tests/test_bonded.py [hunk 3]
```diff
@@ -278,8 +319,8 @@ def test_harmonic_angle(self, n_particles=64, n_angles=25, dim=3):
             test_du_dx_rev, test_du_dp_rev, test_u_rev = test_potential_rev_impl.execute(x, params, box, 1, 1, 1)
 
             np.testing.assert_array_equal(test_u, test_u_rev)
-            np.testing.assert_array_equal(test_du_dx, test_du_dx_rev)
             np.testing.assert_array_equal(test_du_dp, test_du_dp_rev)
+            np.testing.assert_array_equal(test_du_dx, test_du_dx_rev)  # we lose this when we use the cross product
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:323]: Sorry - this *was* failing before, but it's been fixed now. Will address.

File: tests/test_bonded.py [hunk 4]
```diff
@@ -244,9 +247,43 @@ def test_harmonic_bond_singularity(self):
...
+            box = np.eye(3) * 100
+            test_u = harmonic_angle(x, params, box, angle_idxs)
+            ref_u = original_harmonic_angle(x, params, box, angle_idxs)
+            np.testing.assert_almost_equal(test_u, ref_u)
+
     def test_harmonic_angle(self, n_particles=64, n_angles=25, dim=3):
+        # def test_harmonic_angle(self, n_particles=3, n_angles=1, dim=3):
         """Randomly connect triples of particles, then validate the resulting HarmonicAngle force"""
         np.random.seed(125)
+        # np.random.seed(126)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:286]: (Spurious comment - had a buggy implementation, will remove comment)

[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:286]: (There's a for loop in the test that's equivalent to randomizing seeds):

`for _ in range(25):`

File: tests/test_bonded.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:323]: (We no longer use the cross-product in the implementation).
